Configuration-Interaction/helper_CI.py

```diff
@@ -0,0 +1,366 @@
+        
+        for i1,i2 in combinations(alphaO,2):
+            for j1,j2 in combinations(alphaU,2):
+                det = self.copy()
+                det.removeAlphaOrbital(i1)
+                det.addAlphaOrbital(j1)
+                det.removeAlphaOrbital(i2)
+                det.addAlphaOrbital(j2)
+                dets.append(det)
+        
+        for k1,k2 in combinations(betaO,2):
+            for l1,l2 in combinations(betaU,2):
+                det = self.copy()
+                det.removeBetaOrbital(k1)
+                det.addBetaOrbital(l1)
+                det.removeBetaOrbital(k2)
+                det.addBetaOrbital(l2)
+                dets.append(det)
+        return dets
+    
+    def generateSingleAndDoubleExcitationsOfDet(self, nmo):
+        """generate all the single and double excitations of determinant in a list"""
+        return self.generateSingleExcitationsOfDet(nmo) + self.generateDoubleExcitationsOfDet(nmo)
+    
+    def copy(self):
+        """Return a deep copy of self"""
+        return Determinant(alphaObtBits=self.alphaObtBits, betaObtBits=self.betaObtBits)
+    
+    def __str__(self):
+        a,b = self.getOrbitalIndexLists()
+        return "|"+str(a)+str(b)+">"
+
+
+import numpy as np
+
+class HamiltonianGenerator:
+    '''
+    class for Full CI matrix elements
+    '''
+    def __init__(self, H_spin, mo_spin_eri):
+        """Constructor for MatrixElements"""
+        self.Hspin = H_spin
+        self.antiSym2eInt = mo_spin_eri
+    
+    def generateMatrix(self, detList):
+        """Generate CI Matrix"""
+        numDet = len(detList)
+        matrix = np.zeros((numDet,numDet))
+        for i in xrange(numDet):
+            for j in xrange(i+1):
+                matrix[i,j] = self.calcMatrixElement(detList[i], detList[j])
+                matrix[j,i] = matrix[i,j]
+        return matrix
+    
+    def calcMatrixElement(self, det1, det2):
+        """Calculate a matrix element by two determinants"""
+        numUniqueOrbitals = None
+        if det1.diff2OrLessOrbitals(det2):
+            numUniqueOrbitals = det1.numberOfTotalDiffOrbitals(det2)
+            if numUniqueOrbitals == 2:
+                return self.calcMatrixElementDiffIn2(det1, det2)
+            elif numUniqueOrbitals == 1:
+                return self.calcMatrixElementDiffIn1(det1, det2)
+            else:
+                return self.calcMatrixElementIdentialDet(det1)
+        else:
+            return 0.0
+    
+    def calcMatrixElementDiffIn2(self, det1, det2):
+        """Calculate a matrix element by two determinants where the determinants differ by 2 spin orbitals"""
+        unique1, unique2, sign = det1.getUniqueOrbitalsInMixIndexListsPlusSign(det2)
+        return sign * self.antiSym2eInt[unique1[0], unique1[1], unique2[0], unique2[1]]
+    
+    def calcMatrixElementDiffIn1(self, det1, det2):
+        """Calculate a matrix element by two determinants where the determinants differ by 1 spin orbitals"""
```

Comment:
For each of these could you write the matrix forms of the slater rules? `H[m, n] = sign * ( h_{m,n} + <m p || n p >)` or the like.